hazelcast/src/main/java/com/hazelcast/cp/internal/datastructures/spi/blocking/AbstractBlockingService.java

```diff
@@ -366,6 +367,9 @@ public void run() {
         @Override
         public void run() {
             for (Entry<CPGroupId, Collection<BiTuple<String, UUID>>> e : getWaitKeysToExpire().entrySet()) {
+                if (currentThread().isInterrupted()) {
+                    break;
+                }
                 tryReplicateExpiredWaitKeys(e.getKey(), e.getValue());
```

Comment:
If this thread is blocked inside `tryReplicateExpiredWaitKeys()` method, I don't think this change will help. I don't remember how `raftNode.replicate(op).get()` behaves when the local member is shutdown. We should investigate it too.

Reply:
I have added timeout there as you suggested.